### PR TITLE
Make `Sized` coinductive, again

### PR DESCRIPTION
**Pull request by compiler-errors**
A revival of #83647

---

What exactly makes co-induction sound? Better question: are there any unsoundness risks from this? `Sized` can't be implemented by custom `impl` blocks, nor can it be conditionally implemented based on anything other than child fields being `Sized`, right?

r? @nikomatsakis for whenever he gets back from vacation

### CODE REVIEW

**Comment by nikomatsakis** [comment]
I think we should do this, but I think it requires an FCP. 

The soundness challenges with coinduction are tied to elaborated bounds, which don't apply to `Sized` as it has no supertraits.

Ideally we'd have an RFC for moving all trait matching to coinduction.

**Comment by compiler-errors** [comment]
Since exploring moving all traits to be coinductive is significantly harder and likely requires an RFC/opt-in/migration/etc., I'm following Nikos'a advice above and proposing merging just this for now.

@rfcbot merge

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by madsmtm** [comment]
> don't apply to `Sized` as it has no supertraits.

Just wanted to note that it is being considered in https://github.com/rust-lang/rust/issues/81513 to add `Thin` as a supertrait of `Sized`.

I don't know anything about coinduction, but if the only reason this is sound is that `Sized` has no supertraits, then you may want to reconsider doing this?

**Comment by compiler-errors** [comment]
@madsmtm:

I'm not sure if it makes sense to add `Thin` as a supertrait of `Sized` nowadays -- the `Pointee` trait is implemented for all types, but there is [special logic](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_trait_selection/traits/project.rs.html#1688) in the compiler to make sure that `T: Sized => T: Thin` without needing to enforce that relationship as a supertrait (#93977).

I'm not aware of any benefits of that relationship being reflected as a supertrait relationship, other than simplifying some of the builtin trait code... I can put that in as a concern and raise it up at a T-types meeting though.

**Comment by compiler-errors** [comment]
@rfcbot concern does this limit Sized having supertrait predicates (like `trait Thin =Pointee<Metadata = ()>`) in the future?

**Comment by compiler-errors** [comment]
(Presumably not -- we could work around this with the same modifications to the trait system as we need to enable coinduction for _all_ traits, which we eventually do want. But it would be a shame to block a library feature that many people want on a trait system modification.)

**Comment by Jules-Bertholet** [comment]
https://github.com/rust-lang/rfcs/pull/3319 would add `Aligned` as a supertrait of `Sized`.

**Comment by nikomatsakis** [comment]
In general, having supertrait predicates is not a concern, but we will have to do some work on our trait system implementation before we can properly handle it.

**Comment by compiler-errors** [comment]
@nikomatsakis, do you think we should block this due to possible RFCs adding supertrait bounds to `Sized`? I'm not sure if I have a strong opinion on this.

**Comment by lcnr** [comment]
I don't think we should block this on adding a super trait. Making `Sized` coinductive rn has a noticeable positive impact as it is necessary to abstract over collections for recursive structures, see src/test/ui/sized/coinductive-2.rs added in this PR.

We intend to support super traits for coinductive types in the future anyways and I expect us to be able to deal with this once necessary.

**Comment by compiler-errors** [comment]
@rfcbot resolve does this limit Sized having supertrait predicates (like `trait Thin =Pointee<Metadata = ()>`) in the future?

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 1]
```diff
@@ -956,7 +956,10 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
 
     fn coinductive_predicate(&self, predicate: ty::Predicate<'tcx>) -> bool {
         let result = match predicate.kind().skip_binder() {
-            ty::PredicateKind::Trait(ref data) => self.tcx().trait_is_auto(data.def_id()),
+            ty::PredicateKind::Trait(ref data) => {
+                self.tcx().trait_is_auto(data.def_id())
+                    || self.tcx().lang_items().sized_trait() == Some(data.def_id())
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:961]: can you move that into a function: `self.tcx().trait_is_coinductive(data.def_id())`?

**Review by lcnr — COMMENTED** [review]
nit, then r=me

**Comment by lcnr** [comment]
Should mention this in the full relnotes. This PR changes `Sized` obligations to be coinductive. This means that if the trait solver encounters a cycle which only includes auto traits and now `Sized`, it returns "Success" instead of an error.

An example would be
```rust
struct Node<C: Trait<Self>>(C::Assoc);

trait Trait<T> {
    type Assoc;
}

impl<T> Trait<T> for Vec<()> {
    type Assoc = Vec<T>;
}

fn main() {
    let _ = Node::<Vec<()>>(Vec::new());
}
```
`Node<Vec<()>>` is `Sized` if `<Vec<()> as Trait<Node<Vec<()>>>::Assoc` is `Sized`. After normalization, this requires `Vec<Node<Vec<()>>` to be `Sized`. `Vec<Node<Vec<()>>` again requires `Node<Vec<()>>` to be `Sized`, resulting in a cycle. As `Sized` is now coinductive, this does not result in an error anymore.

File: src/test/ui/sized/coinductive-1.rs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    type Assoc;
+}
+
+impl<T> Trait<T> for Vec<()> {
+    type Assoc = Vec<T>;
+}
+
+fn main() {
+    let _ = Node::<Vec<()>>(Vec::new());
+}
```
Comment (lcnr) [inline src/test/ui/sized/coinductive-1.rs:14]: can you add the same example but with GATs as well? `trait Trait { type Assoc<T>; }`

**Comment by compiler-errors** [comment]
whoops I forgot to approve this after I rebased

@bors r=lcnr

**Pull request merged**